### DISCUSSION TITLE
Reduce the number of builds by deploying from "release"

### DISCUSSION THREAD

**Discussion by nioc** [sveltia/sveltia-cms#82] [Ideas]
Hello.

Would it be possible to trigger a release (with a tag or something) in order to limit the number of builds generated?
Idea heavily inspired by this: https://github.com/decaporg/decap-cms/pull/6854 😄

**Comment by kyoshino** [comment]
I saw the Decap issue before. Not sure how it really works or if releases are actually needed (I think pull requests are easier to manage), but in general, pushing multiple changes at once makes a lot of sense. I’d call it Shopping Cart mode. Will take a look further to see what’s the best way. Perhaps some other CMSes have better ideas.

Related: I plan to implement https://github.com/decaporg/decap-cms/issues/3821 so an entry (e.g. a blog post) and new relation field values (e.g. tags) can be pushed at once.

**Comment by kyoshino** [comment]
Some of my clients use Cloudflare Pages so I’m familiar with the service that was mentioned in the original Decap issue, but not sure yet how to manage the release-based deployments. If we want to [skip builds](https://developers.cloudflare.com/pages/platform/branch-build-controls/#skip-builds), [adding `[skip ci]`](https://docs.github.com/en/actions/managing-workflow-runs/skipping-workflow-runs) to commit messages is way easier. I could implement it quickly 🙂 

See also: [Deploy hooks](https://developers.cloudflare.com/pages/platform/deploy-hooks)

**Reply by nioc** [reply]
Thank you for your prompt and comprehensive reply. I'm planning to use CF worker for CMS authentication but GitHub Pages for hosting.
My problem concerns the overconsumption of resources for unwanted builds, from a Green IT perspective 😉
I have to figure out how not to trigger the build from the Github CI with the `[skip ci]` feature (I did not know about it), but how writer can tell it from the CMS?

**Reply by kyoshino** [reply]
Oh, I think adding `[skip ci]` is possible with [Commit Message Templates](https://decapcms.org/docs/beta-features/#commit-message-templates), but I could implement the option to the CMS itself 🙂

**Reply by nioc** [reply]
It could be great, may be as a switch near the save button or a check in the drop down:
![Screenshot_20231129_184816_Firefox](https://github.com/sveltia/sveltia-cms/assets/5225411/f2dc8a5b-d153-4da8-9f71-00557946a491)

**Reply by kyoshino** [reply]
Was thinking about the same.

**Reply by kyoshino** [reply]
I’ll work on it this weekend!

**Reply by nioc** [reply]
Great! Thanks.

In parallel, I was looking GH docs for checking how jobs are runned, and I found  [dispatch event API](https://docs.github.com/en/rest/actions/workflows#create-a-workflow-dispatch-event) which could be another option (a new "Publish" button which execute the API request using a configured workflow_id, then the configured workflow build and deploy static on hosting solution).

Of course, this might appears to be a github centric solution and you probably do not want it for this reason.

But I can not ignore it, just saying if it is too complicated or not aligned with your vision 😄

**Reply by kyoshino** [reply]
Thinking about something like this:

- Add a config option to disable auto deployments by default 
    ```yaml
    backend:
      auto_deployments: false
    ```
- If disabled:
    - Append `[skip ci]` to commit messages
    - Show the Deploy button on the global toolbar that can manually trigger GitHub Actions
    - Show a setting where the user can enter an alternative webhook URL, e.g. [Cloudflare Pages](https://developers.cloudflare.com/pages/platform/deploy-hooks) (The URL can’t be hardcoded into the config file because it’s secret)
- Regardless of the config:
    - Provide the “Save” and “Save & Deploy” options on the entry editor’s toolbar
    - Save button will always append `[skip ci]`, Save & Deploy won’t

**Reply by nioc** [reply]
Wow, it looks perfect to me. Thanks for this analysis ❤️

**Comment by kyoshino** [comment]
I have a good progress. Will be testing and shipping it hopefully tonight.

I think having a release management system in a CMS is overkill and unintuitive. If we really want to have it, we should implement [roles](https://github.com/decaporg/decap-cms/issues/2) first to make sure only admins have access to it. I plan to implement roles in Sveltia CMS anyway, which would be a prerequisite for the [config editor](https://github.com/decaporg/decap-cms/issues/341).

**Comment by kyoshino** [comment]
Just shipped [v0.9.0](https://github.com/sveltia/sveltia-cms/releases/tag/v0.9.0) with the fix! 🎉

**Reply by nioc** [reply]
Thanks a lot, it looks perfect!
I'll give a look tomorrow.

**Reply by nioc** [reply]
I was able to test it tonight, and it works like a charm:
https://github.com/nioc/test-hugo/compare/0.1.0...master

Good job and super fast delivery ❤️
